.github/workflows/test-cpu.yml

```diff
@@ -0,0 +1,116 @@
+name: Python application
+
+on:
+  push:
+    branches:
+      - main
+      - "*.*.x"
+  pull_request:
+    branches:
+      - main
+      - "*.*.x"
+# Cancel the job if new commits are pushed
+# https://stackoverflow.com/questions/66335225/how-to-cancel-previous-runs-in-the-pr-when-you-push-new-commitsupdate-the-curre
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  pytest:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        include:
+          - python-version: 3.13
+            test-type: coverage
+          - python-version: 3.11
+            test-type: standard
+          - python-version: 3.13
+            dependencies-version: pre-release
+            test-type: strict-warning
+          - python-version: 3.11
+            dependencies-version: minimum
+            test-type: coverage
+    defaults:
+      run:
+        shell: bash -el {0}
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: Install UV
+        uses: hynek/setup-cached-uv@v2
+        with:
+          cache-dependency-path: pyproject.toml
+
+      - name: Install dependencies
+        if: matrix.dependencies-version == null
+        run: uv pip install --system --compile "anndata[dev,test] @ ." -c ci/constraints.txt
+
+      - name: Install minimum dependencies
+        if: matrix.dependencies-version == 'minimum'
+        run: |
+          uv pip install --system --compile tomli packaging
+
+          deps=$(python3 ci/scripts/min-deps.py pyproject.toml --extra dev test)
+          uv pip install --system --compile $deps pytest-cov "anndata @ ."
+
+      - name: Install dependencies release candidates
+        if: matrix.dependencies-version == 'pre-release'
+        run: uv pip install -v --system --compile --pre "anndata[dev,test] @ ." -c ci/constraints.txt
+
+      - name: Display installed versions
+        run: uv pip list
+
+      - name: Run PyTest
+        if: matrix.test-type == 'standard'
+        run: pytest -n auto
+
+      - name: Run PyTest (coverage)
+        if: matrix.test-type == 'coverage'
+        run: pytest --cov --cov-report=xml --cov-context=test -n auto
+
+      - name: Run PyTest (treat warnings as errors)
+        if: matrix.test-type == 'strict-warning'
+        run: pytest --strict-warnings -n auto
+
+      - uses: codecov/codecov-action@v4
```

Comment:
We have to be careful with the new codecov action: https://github.com/codecov/codecov-cli/issues/648

We need to really check that the coverage numbers are correct before we merge this.

Reply:
I thin I will just use v4 in the meantime, you're right about it not working